backend/apis/leads/list-leads.api.js

```diff
@@ -152,71 +147,130 @@ app.get("/api/leads", (req, res, next) => {
+
+        const leadServices = [];
+        const leadEvents = [];
+
+        if (results[0].length > 0) {
+          for (let j = 0; j < results[0].length; j++) {
+            const leadService = {
+              id: results[0][j].serviceId,
+              servicename: results[0][j].serviceName,
+              programId: results[0][j].programId,
+              programName: results[0][j].programName
+            };
+            leadServices.push(leadService);
+          }
+        }
+
+        if (results[1].length > 0) {
+          for (let j = 0; j < results[1].length; j++) {
+            const leadEvent = {
+              id: results[1][j].eventId,
+              eventName: results[1][j].eventName,
+              eventLocation: results[1][j].eventLocation,
+              eventDate: results[1][j].eventDate
+            };
+            leadEvents.push(leadEvent);
+          }
+        }
+
+        const lead = {
+          id: leadRows[i].leadId,
+          dateOfSignUp: responseDateWithoutTime(leadRows[i].dateOfSignUp),
+          leadStatus:
+            leadRows[i].leadStatus === null ? "active" : leadRows[i].leadStatus,
+          contactStage: {
+            first: leadRows[i].firstContactAttempt,
+            second: leadRows[i].secondContactAttempt,
+            third: leadRows[i].thirdContactAttempt
+          },
+          inactivityReason: leadRows[i].inactivityReason,
+          eventSources: leadEvents,
+          firstName: leadRows[i].firstName,
+          lastName: leadRows[i].lastName,
           fullName: responseFullName(
-            result.modifiedByFirstName,
-            result.modifiedByLastName
+            leadRows[i].firstName,
+            leadRows[i].lastName
           ),
-          timestamp: result.dateModified
+          phone: leadRows[i].phone,
+          smsConsent: responseBoolean(leadRows[i].smsConsent),
+          zip: leadRows[i].zip,
+          age: leadRows[i].age,
+          gender: leadRows[i].gender,
+          leadServices: leadServices,
+          clientId: leadRows[i].clientId,
+          isDeleted: responseBoolean(leadRows[i].isDeleted),
+          createdBy: {
+            userId: leadRows[i].addedBy,
+            firstName: leadRows[i].createdByFirstName,
+            lastName: leadRows[i].createdByLastName,
+            fullName: responseFullName(
+              leadRows[i].createdByFirstName,
+              leadRows[i].createdByLastName
+            ),
+            timestamp: leadRows[i].dateAdded
+          },
+          lastUpdatedBy: {
+            userId: leadRows[i].modifiedBy,
+            firstName: leadRows[i].modifiedByFirstName,
+            lastName: leadRows[i].modifiedByLastName,
+            fullName: responseFullName(
+              leadRows[i].modifiedByFirstName,
+              leadRows[i].modifiedByLastName
+            ),
+            timestamp: leadRows[i].dateModified
+          }
+        };
+
+        leads.push(lead);
+
+        if (leads.length === leadRows.length) {
```

Comment:
When is this ever not true.

Reply:
leads.length won't equal leadRows.length until it has ittereated over the whole array. I kept running into it res.send ing before the data was ready from the for loop. Will using forEach solve that or do I still need this?

Reply:
I didn't realize this is in a forEach. This is not how we should be doing things. We should get the data in one step and then send it in the next.

Reply:
I didn't realize this is in a forEach. This is not how we should be doing things. We should get the data in one step and then send it in the next.